From 75c38d42afb07a7083e3d6935d499ed7c808cf03 Mon Sep 17 00:00:00 2001
From: hhh <3224318821@qq.com>
Date: Fri, 6 Feb 2026 23:22:22 +0800
Subject: [PATCH 02/22] testing burst...


diff --git a/rocket-chip b/rocket-chip
index d0c6b50f..ec4324c8 160000
--- a/rocket-chip
+++ b/rocket-chip
@@ -1 +1 @@
-Subproject commit d0c6b50fdefcdbe121e9788433ea51f7efaf1d32
+Subproject commit ec4324c832feb2229220e9248f00e97a5ffb6657
-- 
2.34.1

