From 751b1edf9023f11317abcdee5310aab36d419cd3 Mon Sep 17 00:00:00 2001
From: hhh <3224318821@qq.com>
Date: Sat, 7 Feb 2026 23:02:14 +0800
Subject: [PATCH 03/21] test 1 for evaluate.


diff --git a/perip/amba/apb_delayer.v b/perip/amba/apb_delayer.v
index a96dba4d..5f555e2d 100644
--- a/perip/amba/apb_delayer.v
+++ b/perip/amba/apb_delayer.v
@@ -27,17 +27,17 @@ module apb_delayer(
     parameter ACCESS = 2'b01;
     parameter WAIT   = 2'b10;
 
-    parameter FREQ_CPU = 975;
+    parameter FREQ_CPU = 1037;
     parameter FREQ_PER = 100;
 
     parameter FACTOR = 10;
-    parameter RATE_FIX = 8960; // (9.75 - 1) * 1024
+    parameter RATE_FIX = 10619; // (freq - 1) * 1024  975 -> 8960, 1037 -> 10619
     
     reg [1:0] state, nstate;
 
     reg [31:0] rdata;
 
-    reg [32:0] delay_counter;
+    reg [63:0] delay_counter;
 
     always @(posedge clock) begin
         if (reset) begin
diff --git a/perip/sdram/core_sdram_axi4/sdram_axi_core.v b/perip/sdram/core_sdram_axi4/sdram_axi_core.v
index ab51f6b6..4984e720 100644
--- a/perip/sdram/core_sdram_axi4/sdram_axi_core.v
+++ b/perip/sdram/core_sdram_axi4/sdram_axi_core.v
@@ -284,13 +284,13 @@ begin
     //-----------------------------------------
     STATE_WRITE0 :
     begin
-    //     next_state_r = STATE_WRITE1;
-    // end
+        next_state_r = STATE_WRITE1;
+    end
     //-----------------------------------------
     // STATE_WRITE1
     //-----------------------------------------
-    // STATE_WRITE1 :
-    // begin
+    STATE_WRITE1 :
+    begin
         next_state_r = STATE_IDLE;
 
         // Another pending write request (with no refresh pending)
@@ -630,19 +630,19 @@ begin
     //-----------------------------------------
     // STATE_WRITE1
     //-----------------------------------------
-    // STATE_WRITE1 
-    // begin
-    //     // Burst continuation
-    //     command_q   <= CMD_NOP;
+    STATE_WRITE1 :
+    begin
+        // Burst continuation
+        command_q   <= CMD_NOP;
 
-    //     data_q      <= data_buffer_q;
+        // data_q      <= data_buffer_q;
 
-    //     // Disable auto precharge (auto close of row)
-    //     addr_q[AUTO_PRECHARGE]  <= 1'b0;
+        // Disable auto precharge (auto close of row)
+        // addr_q[AUTO_PRECHARGE]  <= 1'b0;
 
-    //     // Write mask
-    //     dqm_q       <= dqm_buffer_q;
-    // end
+        // Write mask
+        dqm_q       <= 4'b1111;
+    end
     endcase
 end
 
@@ -685,8 +685,8 @@ if (rst_i)
     ack_q   <= 1'b0;
 else
 begin
-    // if (state_q == STATE_WRITE1)
-    if (state_q == STATE_WRITE0)
+    if (state_q == STATE_WRITE1)
+    // if (state_q == STATE_WRITE0)
         ack_q <= 1'b1;
     // else if (rd_q[SDRAM_READ_LATENCY+1])
     else if (rd_q[SDRAM_READ_LATENCY])
diff --git a/perip/sdram/sdram.v b/perip/sdram/sdram.v
index 40238729..43b12bbb 100644
--- a/perip/sdram/sdram.v
+++ b/perip/sdram/sdram.v
@@ -179,6 +179,7 @@ module sdram
                     if(cas_cnt < cas_len - 1) begin
                         cas_cnt <= cas_cnt + 1;
                     end else if(cas_cnt >= cas_len - 1) begin
+                        cas_cnt <= 0;
                         if(burst_cnt < burst_len) begin
                             burst_cnt <= burst_cnt + 1;
                         end
@@ -202,8 +203,8 @@ import "DPI-C" function void sdram_read(input int addr, output int data, input i
 import "DPI-C" function void sdram_write(input int addr, input int data, input int mask, input int num);
 
     wire [SDRAM_ROW_W+SDRAM_COL_W+SDRAM_BANK_W+1:0] full_addr_active, full_addr;
-    assign full_addr_active = {row_active[bank_active], bank_active, col_active[SDRAM_COL_W-1:0], burst_cnt[0], 1'b0};
-    assign full_addr = {row_active[ba], ba, col_addr[SDRAM_COL_W-1:0], burst_cnt[0], 1'b0};
+    assign full_addr_active = {row_active[bank_active], bank_active, col_active[SDRAM_COL_W-1:0], 2'b0};
+    assign full_addr = {row_active[ba], ba, col_addr[SDRAM_COL_W-1:0], 2'b0};
 
     always @(posedge clk) begin
         if(cas_cnt == cas_len - 1 && burst_cnt < burst_len) begin
diff --git a/src/Top.scala b/src/Top.scala
index 2913155b..c00378a6 100644
--- a/src/Top.scala
+++ b/src/Top.scala
@@ -7,7 +7,7 @@ import freechips.rocketchip.diplomacy.LazyModule
 
 object Config {
   def hasChipLink: Boolean = false
-  def sdramUseAXI: Boolean = true
+  def sdramUseAXI: Boolean = false
 }
 
 class ysyxSoCTop extends Module {
-- 
2.34.1

