From fc34523416ceb368c4c1f4938cc1eae610944a39 Mon Sep 17 00:00:00 2001
From: 211180120-Yu Bingyao <211180120@smail.nju.edu.cn>
Date: Thu, 27 Mar 2025 19:02:18 +0800
Subject: [PATCH 09/22] github_makefile_test

---
 perip/ps2/ps2_top_apb.v | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/perip/ps2/ps2_top_apb.v b/perip/ps2/ps2_top_apb.v
index 06c9e21..58bb409 100644
--- a/perip/ps2/ps2_top_apb.v
+++ b/perip/ps2/ps2_top_apb.v
@@ -60,7 +60,7 @@ module ps2_top_apb(
       end
     end
   end
-
+//
 endmodule
 
 
-- 
2.34.1

