From ac4e609cfb03c4f1e0ba0689ab792b9b0f44dee2 Mon Sep 17 00:00:00 2001
From: hhh <3224318821@qq.com>
Date: Tue, 24 Feb 2026 12:30:24 +0800
Subject: [PATCH 10/22] evaluate 6, forward


diff --git a/perip/amba/apb_delayer.v b/perip/amba/apb_delayer.v
index a6b2b872..6d3253f0 100644
--- a/perip/amba/apb_delayer.v
+++ b/perip/amba/apb_delayer.v
@@ -37,8 +37,9 @@ module apb_delayer(
         1008  -> 9298
         858   -> 7762
         869   -> 7875
+        807   -> 7261
     */
-    parameter RATE_FIX = 7875;
+    parameter RATE_FIX = 7261;
     
     reg [1:0] state, nstate;
 
diff --git a/perip/amba/axi4_delayer.v b/perip/amba/axi4_delayer.v
index 988ba3aa..6b65b00f 100644
--- a/perip/amba/axi4_delayer.v
+++ b/perip/amba/axi4_delayer.v
@@ -84,8 +84,9 @@ module axi4_delayer(
         1008  -> 9298
         858   -> 7762
         869   -> 7875
+        807   -> 7261
     */
-    parameter RATE_FIX = 7875;
+    parameter RATE_FIX = 7261;
     
     reg [3:0] state, nstate;
 
-- 
2.34.1

