From 8f1824c76f4258306fff0e61fc7049953da57064 Mon Sep 17 00:00:00 2001
From: WuShFeng <1481542813@qq.com>
Date: Thu, 23 Jul 2026 15:42:10 +0800
Subject: [PATCH 13/24] =?UTF-8?q?feat:=20sdram=E6=94=B9=E7=94=A8axi?=
 =?UTF-8?q?=E5=AE=9E=E7=8E=B0?=
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

---
 src/Top.scala | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/Top.scala b/src/Top.scala
index c00378a6..2913155b 100644
--- a/src/Top.scala
+++ b/src/Top.scala
@@ -7,7 +7,7 @@ import freechips.rocketchip.diplomacy.LazyModule
 
 object Config {
   def hasChipLink: Boolean = false
-  def sdramUseAXI: Boolean = false
+  def sdramUseAXI: Boolean = true
 }
 
 class ysyxSoCTop extends Module {
-- 
2.53.0

