From 417872e9027f95d439b1953abcf5f77fb87330c4 Mon Sep 17 00:00:00 2001
From: Shitful <2292683261@qq.com>
Date: Mon, 8 Jun 2026 00:07:18 +0800
Subject: [PATCH 13/46] dummy 9

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 84e5aa10..c01e506e 100644
--- a/Makefile
+++ b/Makefile
@@ -11,7 +11,7 @@ $(V_FILE_FINAL): $(SCALA_FILES)
 # TODO: This can be removed after chisel publishes a new version
 	sed -i  "s/RegNext(r_sel0)/Reg(Bool())/g" rocket-chip/src/main/scala/amba/axi4/SRAM.scala || true
 	sed -i  "s/RegNext(w_sel0)/Reg(Bool())/g" rocket-chip/src/main/scala/amba/axi4/SRAM.scala || true
-#	dummy 8
+#	dummy 9
 	@./patch/update-firtool.sh $(FIRTOOL_VERSION) $(FIRTOOL_PATCH_DIR)
 	CHISEL_FIRTOOL_PATH=$(FIRTOOL_PATCH_DIR)/firtool-$(FIRTOOL_VERSION)/bin \
 	mill -i ysyxsoc.runMain ysyx.Elaborate --target-dir $(@D)
-- 
2.55.0

