From 7b930f78f1fb84abe14613844f03be69e3787835 Mon Sep 17 00:00:00 2001
From: kuikuikuizzZ <kimmmywu@gmail.com>
Date: Fri, 22 Aug 2025 19:22:00 +0800
Subject: [PATCH 17/23] feat(soc): add axi delayer

---
 src/amba/AXI4Delayer.scala | 81 +++++++++++++++++++++++++++++++++++++-
 1 file changed, 79 insertions(+), 2 deletions(-)

diff --git a/src/amba/AXI4Delayer.scala b/src/amba/AXI4Delayer.scala
index b60bb57..82a12d5 100644
--- a/src/amba/AXI4Delayer.scala
+++ b/src/amba/AXI4Delayer.scala
@@ -22,7 +22,84 @@ class axi4_delayer extends BlackBox {
 
 class AXI4DelayerChisel extends Module {
   val io = IO(new AXI4DelayerIO)
-  io.out <> io.in
+
+
+  val s_idle :: s_transmit :: s_wait :: s_resp :: Nil = Enum(4)
+  val state = RegInit(s_idle)
+  // B2 freq 350MHz, device typical 100MHz
+  // s = 2^16, r= 3.5 , s_r = 229376
+  val s_r = 229376.U(32.W)
+  val cnt = RegInit(0.U(32.W))
+  val axi_start = io.in.ar.valid || io.in.aw.valid
+  val axi_finsh = (io.out.r.valid &&  io.out.r.ready) || (io.out.b.valid && io.out.b.ready)
+  val is_read = RegEnable(io.in.ar.valid,io.in.ar.valid || state === s_idle)
+  val is_write = RegEnable(io.in.aw.valid,io.in.aw.valid || state === s_idle)
+
+  val rvalid   =  RegEnable(io.out.r.valid ,io.out.r.valid || state === s_idle)
+  val rdata    =  RegEnable(io.out.r.bits.data  ,io.out.r.valid && state === s_transmit)
+  val rresp    =  RegEnable(io.out.r.bits.resp  ,io.out.r.valid && state === s_transmit)
+  val rid      =  RegEnable(io.out.r.bits.id    ,io.out.r.valid && state === s_transmit)
+  val rlast    =  RegEnable(io.out.r.bits.last  ,io.out.r.valid && state === s_transmit)
+
+  
+  val bvalid  = RegEnable(io.out.b.valid ,io.out.b.valid || state === s_idle)
+  val bresp   = RegEnable(io.out.b.bits.resp  ,io.out.b.valid)
+  val bid     = RegEnable(io.out.b.bits.id    ,io.out.b.valid)
+
+
+  switch (state) {
+    is (s_idle) { when (axi_start) { state := s_transmit
+                    cnt := s_r } }
+    is (s_transmit) { when (axi_finsh){ 
+                        state := s_wait
+                        cnt := cnt >> 16
+                      }.otherwise { cnt := cnt + s_r }}
+    is (s_wait) { when (cnt === 1.U) { state := s_resp
+                  }.otherwise { cnt := cnt - 1.U }}
+    is (s_resp) {
+      when(is_read && rvalid && !rlast ){
+        state := s_transmit
+        cnt := s_r
+      } .otherwise {
+        state := s_idle
+      }
+    }
+  }
+
+  io.out := DontCare
+  io.out.ar.valid   := Mux(state===s_idle || state===s_transmit,io.in.ar.valid,false.B)
+  io.out.ar.bits.addr    := io.in.ar.bits.addr 
+  io.out.ar.bits.id      := io.in.ar.bits.id   
+  io.out.ar.bits.len     := io.in.ar.bits.len  
+  io.out.ar.bits.size    := io.in.ar.bits.size 
+  io.out.ar.bits.burst   := io.in.ar.bits.burst
+  io.in.ar.ready    := io.out.ar.ready
+
+  io.out.aw.valid   := Mux(state===s_idle || state===s_transmit,io.in.aw.valid,false.B)
+  io.out.aw.bits.addr    := io.in.aw.bits.addr  
+  io.out.aw.bits.id      := io.in.aw.bits.id    
+  io.out.aw.bits.len     := io.in.aw.bits.len   
+  io.out.aw.bits.size    := io.in.aw.bits.size  
+  io.out.aw.bits.burst   := io.in.aw.bits.burst 
+  io.in.aw.ready    := io.out.aw.ready 
+
+  io.out.w.valid    := Mux(state===s_idle || state===s_transmit,io.in.w.valid,false.B)
+  io.out.w.bits.data     := io.in.w.bits.data    
+  io.out.w.bits.strb     := io.in.w.bits.strb    
+  io.out.w.bits.last     := io.in.w.bits.last    
+  io.in.w.ready     := io.out.w.ready   
+
+  io.out.b.ready    := Mux(state===s_idle || state===s_transmit,io.in.b.ready,false.B)
+  io.in.b.valid     := Mux(state === s_resp, bvalid, false.B)  
+  io.in.b.bits.resp      := bresp   
+  io.in.b.bits.id        := bid  
+  
+  io.out.r.ready    := Mux(state===s_transmit,true.B,false.B)
+  io.in.r.valid     := Mux(state === s_resp, rvalid, false.B)
+  io.in.r.bits.data      := rdata 
+  io.in.r.bits.resp      := rresp 
+  io.in.r.bits.id        := rid   
+  io.in.r.bits.last      := rlast 
 }
 
 class AXI4DelayerWrapper(implicit p: Parameters) extends LazyModule {
@@ -31,7 +108,7 @@ class AXI4DelayerWrapper(implicit p: Parameters) extends LazyModule {
   lazy val module = new Impl
   class Impl extends LazyModuleImp(this) {
     (node.in zip node.out) foreach { case ((in, edgeIn), (out, edgeOut)) =>
-      val delayer = Module(new axi4_delayer)
+      val delayer = Module(new AXI4DelayerChisel)
       delayer.io.clock := clock
       delayer.io.reset := reset
       delayer.io.in <> in
-- 
2.43.0

