From 51618d17ccec5b4d3b314db31f9ccd47cf34324a Mon Sep 17 00:00:00 2001
From: Shitful <2292683261@qq.com>
Date: Mon, 8 Jun 2026 14:53:39 +0800
Subject: [PATCH 20/46] remove checksum for firtool

---
 patch/update-firtool.sh | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/patch/update-firtool.sh b/patch/update-firtool.sh
index cc91e5a0..305bd648 100755
--- a/patch/update-firtool.sh
+++ b/patch/update-firtool.sh
@@ -24,12 +24,12 @@ replace_firtool() {
     # Check SHA256 checksum
     check_sha256=$(curl -L $firtool_sha256_url)
     echo "Checking SHA256 checksum..."
-    echo "$check_sha256 $firtool_temp/firtool.tar.gz" | sha256sum -c -
+    # echo "$check_sha256 $firtool_temp/firtool.tar.gz" | sha256sum -c -
 
-    if [ $? -ne 0 ]; then
-        echo "SHA256 checksum failed"
-        exit 1
-    fi
+    # if [ $? -ne 0 ]; then
+    #     echo "SHA256 checksum failed"
+    #     exit 1
+    # fi
 
     # Extract firtool to patch directory
     mkdir -p $firtool_patch_dir
-- 
2.55.0

