From ec4d1e43ef85d67602b1f8750e4f83ec1c019d4e Mon Sep 17 00:00:00 2001
From: kuikuikuizzZ <kimmmywu@gmail.com>
Date: Thu, 25 Sep 2025 21:37:09 +0800
Subject: [PATCH 20/23] chore(soc): Makefile build in soc_home

---
 Makefile | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/Makefile b/Makefile
index 4b93df3..ee32676 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 V_FILE_GEN   = build/ysyxSoCTop.sv
 V_FILE_FINAL = build/ysyxSoCFull.v
 SCALA_FILES = $(shell find src/ -name "*.scala")
-
+SOC_HOME = $(YSYX_HOME)/ysyxSoC
 # Firtool version
 FIRTOOL_VERSION = 1.105.0
 FIRTOOL_PATCH_DIR = $(shell pwd)/patch/firtool
@@ -19,7 +19,7 @@ $(V_FILE_FINAL): $(SCALA_FILES)
 
 verilog: $(V_FILE_FINAL)
 # 	cp $(V_FILE_FINAL) $(NPC_HOME)/build/
-# 	cp -r $(wildcard $(SOC_HOME)/perip/*/*.v $(SOC_HOME)/perip/psram/efabless/*.v $(SOC_HOME)/perip/sdram/core_sdram_axi4/*.v )  $(NPC_HOME)/build/
+	cp -r $(wildcard $(SOC_HOME)/perip/*/*.v $(SOC_HOME)/perip/psram/efabless/*.v $(SOC_HOME)/perip/sdram/core_sdram_axi4/*.v)  $(SOC_HOME)/build/
 
 clean:
 	-rm -rf build/
-- 
2.43.0

