From 3619eecd38e3c9dc58c7710765e58f5c6b8abfdf Mon Sep 17 00:00:00 2001
From: dengzibin <1447762756@qq.com>
Date: Wed, 29 Apr 2026 15:14:07 +0800
Subject: [PATCH 22/27] cache_mini

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 7a5fc575..f5c0e984 100644
--- a/Makefile
+++ b/Makefile
@@ -45,7 +45,7 @@ CSRCSS += $(SRC_AUTO_BIND)
 CHEAD = $(YSYX_HOME)/ysyxSoC/csrc/include
 
 SDC_FILE = $(YSYX_HOME)/ysyxSoC/sdc/cpu_top.sdc
-RESULT_DIR = $(shell find $(abspath /sta_result))
+RESULT_DIR = $(YSYX_HOME)/ysyxSoC/sta_result
 DESIGN = ysyx_25040102
 
 # rules for NVBoard
-- 
2.34.1

