From a4b553450fa2b38264d18862292c4eb43f15c728 Mon Sep 17 00:00:00 2001
From: dengzibin <1447762756@qq.com>
Date: Thu, 30 Apr 2026 12:41:44 +0800
Subject: [PATCH 23/27] delay4

---
 perip/amba/axi4_delayer.v | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/perip/amba/axi4_delayer.v b/perip/amba/axi4_delayer.v
index a1432152..e92ebb74 100644
--- a/perip/amba/axi4_delayer.v
+++ b/perip/amba/axi4_delayer.v
@@ -100,7 +100,7 @@ module axi4_delayer(
   reg   [31:0]    total_wr;
   reg   [31:0]    counter_wr;
 
-  localparam  R   = 8;
+  localparam  R   = 4;
 
   localparam  IDLE_READ       = 0;
   localparam  WAIT_OUT_RREADY = 2;
-- 
2.34.1

