From cc6c04e5a7928d38ad1ae482b90c6e342cfc60f5 Mon Sep 17 00:00:00 2001
From: Wang Yi Jie <miwyj@mail.scut.edu.cn>
Date: Mon, 2 Mar 2026 22:27:02 +0800
Subject: [PATCH 28/28] set default delay to 1

---
 src/Top.scala | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/Top.scala b/src/Top.scala
index 84e7d8a4..0488090b 100644
--- a/src/Top.scala
+++ b/src/Top.scala
@@ -8,7 +8,7 @@ import freechips.rocketchip.diplomacy.LazyModule
 object Config {
   def hasChipLink: Boolean = false
   def sdramUseAXI: Boolean = true
-  def DelayRatio: Float = sys.env.getOrElse("DELAY_RATIO", "2").toFloat
+  def DelayRatio: Float = sys.env.getOrElse("DELAY_RATIO", "1").toFloat
 }
 
 class ysyxSoCTop extends Module {
-- 
2.43.0

