From 80db3e55bed34b73f7e0cf2caf7bb8ca95cf7cca Mon Sep 17 00:00:00 2001
From: Shitful <2292683261@qq.com>
Date: Wed, 10 Jun 2026 22:09:44 +0800
Subject: [PATCH 37/46] dm

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 270055cd..78cdc1ae 100644
--- a/Makefile
+++ b/Makefile
@@ -11,7 +11,7 @@ $(V_FILE_FINAL): $(SCALA_FILES)
 # TODO: This can be removed after chisel publishes a new version
 	sed -i  "s/RegNext(r_sel0)/Reg(Bool())/g" rocket-chip/src/main/scala/amba/axi4/SRAM.scala || true
 	sed -i  "s/RegNext(w_sel0)/Reg(Bool())/g" rocket-chip/src/main/scala/amba/axi4/SRAM.scala || true
-#	dummy 28
+#	dummy 29
 	@./patch/update-firtool.sh $(FIRTOOL_VERSION) $(FIRTOOL_PATCH_DIR)
 	CHISEL_FIRTOOL_PATH=$(FIRTOOL_PATCH_DIR)/firtool-$(FIRTOOL_VERSION)/bin \
 	mill -i ysyxsoc.runMain ysyx.Elaborate --target-dir $(@D)
-- 
2.55.0

