From 069124ec5658bced6566978ab6b317cbf2a01a24 Mon Sep 17 00:00:00 2001
From: huhao <huhao1226968736@outlook.com>
Date: Wed, 22 Oct 2025 16:52:31 +0800
Subject: [PATCH 2/2] for ci patch

---
 bsp/abstract-machine/src/uart.c | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/bsp/abstract-machine/src/uart.c b/bsp/abstract-machine/src/uart.c
index 46e115a5e..ea76e96f9 100644
--- a/bsp/abstract-machine/src/uart.c
+++ b/bsp/abstract-machine/src/uart.c
@@ -38,12 +38,12 @@ static int _uart_putc(struct rt_serial_device *serial, char c) {
 }
 
 static int _uart_getc(struct rt_serial_device *serial) {
-  static const char *p = "help\ndate\nversion\nfree\nps\npwd\nls\nmemtrace\nmemcheck\nutest_list\n";
-  // char ch = io_read(AM_UART_RX).data;
-  // if (ch == (char)-1) return -1;
-  // if (ch != 8) putch(ch);
-  // return ch;
-  return (*p != '\0' ? *(p ++) : -1);
+  // static const char *p = "help\ndate\nversion\nfree\nps\npwd\nls\nmemtrace\nmemcheck\nutest_list\n";
+  char ch = io_read(AM_UART_RX).data;
+  if (ch == (char)-1) return -1;
+  if (ch != 8) putch(ch);
+  return ch;
+  // return (*p != '\0' ? *(p ++) : -1);
 }
 
 const struct rt_uart_ops _uart_ops = {
-- 
2.34.1

