From 87764df172304d171fe84cf83c78963a7ded515f Mon Sep 17 00:00:00 2001
From: Furina318 <1145676148@qq.com>
Date: Thu, 18 Sep 2025 20:29:36 +0800
Subject: [PATCH 2/6] =?UTF-8?q?rtt=E8=A1=A5=E4=B8=811?=
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

---
 bsp/abstract-machine/src/uart.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/bsp/abstract-machine/src/uart.c b/bsp/abstract-machine/src/uart.c
index a1d6b9dd6..156a8cdf2 100644
--- a/bsp/abstract-machine/src/uart.c
+++ b/bsp/abstract-machine/src/uart.c
@@ -37,7 +37,7 @@ static int _uart_putc(struct rt_serial_device *serial, char c) {
 }
 
 static int _uart_getc(struct rt_serial_device *serial) {
-  static const char *p = "help\ndate\nversion\nfree\nps\npwd\nls\nmemtrace\nmemcheck\nutest_list\nam_hello\nam_microbench test\n";
+  static const char *p = "help\ndate\nversion\nfree\nps\npwd\nls\nmemtrace\nmemcheck\nutest_list\nam_hello\n";
   return (*p != '\0' ? *(p ++) : -1);
 }
 
-- 
2.34.1

