From 0b7f5cdd6fa98d648aa6b4d8c338406dd1dc1457 Mon Sep 17 00:00:00 2001
From: Water Block <17048545@qq.com>
Date: Wed, 20 May 2026 13:45:30 +0800
Subject: [PATCH 2/2] Add Wno-error

---
 bsp/abstract-machine/Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/bsp/abstract-machine/Makefile b/bsp/abstract-machine/Makefile
index 3013b74c8..e3afbb9f3 100644
--- a/bsp/abstract-machine/Makefile
+++ b/bsp/abstract-machine/Makefile
@@ -8,7 +8,7 @@ NAME = rtthread
 SRCS = $(shell find src -name "*.c")
 LDSCRIPTS += extra.ld
 CFLAGS += -DHAVE_CCONFIG_H -D__RTTHREAD__
-CFLAGS += -Wno-nonnull-compare
+CFLAGS += -Wno-error=stringop-overread -Wno-error=nonnull-compare -Wno-error=stringop-overflow
 -include $(FILE_MK)
 -include $(AM_APPS_MK)
 include $(AM_HOME)/Makefile
-- 
2.43.0

