From 82e379d95fc9501c23d225f4dd586841a664f297 Mon Sep 17 00:00:00 2001
From: kuikuikuizzZ <kimmmywu@gmail.com>
Date: Thu, 25 Sep 2025 15:59:22 +0800
Subject: [PATCH 2/5] feat(bsp): update Makefile

---
 bsp/abstract-machine/Makefile | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/bsp/abstract-machine/Makefile b/bsp/abstract-machine/Makefile
index 8d70f310b..b45a171b0 100644
--- a/bsp/abstract-machine/Makefile
+++ b/bsp/abstract-machine/Makefile
@@ -8,7 +8,9 @@ NAME = rtthread
 SRCS = $(shell find src -name "*.c")
 CFLAGS += -DHAVE_CCONFIG_H -D__RTTHREAD__
 CFLAGS += -Wno-nonnull-compare
-LDFLAGS += -T extra.ld
+ifneq ($(ARCH),riscv32e-ysyxsoc)
+	LDFLAGS += -T extra.ld
+endif
 -include $(FILE_MK)
 -include $(AM_APPS_MK)
 include $(AM_HOME)/Makefile
-- 
2.43.0

