From 440a8ffba324efae986cd0dd630fb8ba7c276ba2 Mon Sep 17 00:00:00 2001
From: PengchengYang-xdu <24241214835@stu.xidian.edu.cn>
Date: Wed, 8 Oct 2025 22:26:32 +0800
Subject: [PATCH 3/9] fix rtt uart include

---
 bsp/abstract-machine/src/uart.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/bsp/abstract-machine/src/uart.c b/bsp/abstract-machine/src/uart.c
index 939714e0f..485d9076b 100755
--- a/bsp/abstract-machine/src/uart.c
+++ b/bsp/abstract-machine/src/uart.c
@@ -11,6 +11,7 @@
 #include <rtthread.h>
 #include <am.h>
 #include <klib.h>
+#include <klib-macros.h>
 
 #define UART_DEFAULT_BAUDRATE 115200
 
-- 
2.34.1

