From 8c4c4bcdccd41b2b65fc3a6ab558b5015e14fc64 Mon Sep 17 00:00:00 2001
From: EPTansuo <eptansuo@163.com>
Date: Thu, 21 Nov 2024 18:42:12 +0800
Subject: [PATCH 4/8] modified gitignore

---
 bsp/abstract-machine/.gitignore | 1 +
 1 file changed, 1 insertion(+)

diff --git a/bsp/abstract-machine/.gitignore b/bsp/abstract-machine/.gitignore
index 27ab723ff8..f903933d3c 100644
--- a/bsp/abstract-machine/.gitignore
+++ b/bsp/abstract-machine/.gitignore
@@ -1,2 +1,3 @@
 rtconfig.h
 files.mk
+*.tar.gz
-- 
2.50.1

