From f0678f19125c508d9355c2f29915fa851a224fe2 Mon Sep 17 00:00:00 2001
From: dengzibin <1447762756@qq.com>
Date: Mon, 27 Apr 2026 19:22:01 +0800
Subject: [PATCH 4/7] remove_txt

---
 .gitignore | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/.gitignore b/.gitignore
index 558b09f..177ff6d 100644
--- a/.gitignore
+++ b/.gitignore
@@ -51,3 +51,5 @@ tags
 CMakeLists.txt
 cmake-build-debug
 *.mk
+
+*.patch
\ No newline at end of file
-- 
2.34.1

