From 259571843c7eab22903682df9863424972091b74 Mon Sep 17 00:00:00 2001
From: Furina318 <1145676148@qq.com>
Date: Mon, 29 Sep 2025 22:21:31 +0800
Subject: [PATCH 6/6] =?UTF-8?q?=E8=A1=A5=E4=B8=811?=
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

---
 bsp/abstract-machine/src/uart.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/bsp/abstract-machine/src/uart.c b/bsp/abstract-machine/src/uart.c
index c553a3e6d..08a463d1b 100644
--- a/bsp/abstract-machine/src/uart.c
+++ b/bsp/abstract-machine/src/uart.c
@@ -38,7 +38,7 @@ static int _uart_putc(struct rt_serial_device *serial, char c) {
 }
 
 static int _uart_getc(struct rt_serial_device *serial) {
-  static const char *const pp = "help\ndate\nversion\nfree\nps\npwd\nls\nmemtrace\nmemcheck\nutest_list\nam_hello\nam_microbench test\n";
+  static const char *const pp = "help\ndate\nversion\nfree\nps\npwd\nls\nmemtrace\nmemcheck\nutest_list\n";
   static const char *p = pp;
   static bool exhausted = false;
 
-- 
2.34.1

