From dc6a59348b0951b166d2392fa9c75fd0586fdf9f Mon Sep 17 00:00:00 2001
From: hhh <3224318821@qq.com>
Date: Mon, 9 Mar 2026 20:34:35 +0800
Subject: [PATCH 08/14] am_hello


diff --git a/bsp/abstract-machine/src/uart.c b/bsp/abstract-machine/src/uart.c
index 1b8df7a..e5aa265 100644
--- a/bsp/abstract-machine/src/uart.c
+++ b/bsp/abstract-machine/src/uart.c
@@ -40,7 +40,7 @@ static int _uart_putc(struct rt_serial_device *serial, char c) {
 static int _uart_getc(struct rt_serial_device *serial) {
   // char ch = io_read(AM_UART_RX).data; 
   // return (int)ch;
-  static const char *p = "help\nam_microbench\n";
+  static const char *p = "help\nam_hello\nam_microbench\n";
   return (*p != '\0' ? *(p ++) : -1);
 }
 
-- 
2.34.1

